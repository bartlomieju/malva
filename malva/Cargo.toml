[package]
name = "malva"
version = "0.0.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "CSS/SCSS/Sass/Less formatter."
repository = "https://github.com/g-plane/malva"
license = "MIT"

[dependencies]
itertools = "0.11"
raffia = "0.4"
tiny_pretty = { path = "../../tiny_pretty" }
