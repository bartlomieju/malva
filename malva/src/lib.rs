mod ctx;
mod doc_gen;
pub mod options;
