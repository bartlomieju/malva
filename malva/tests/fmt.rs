use insta::{assert_snapshot, glob, Settings};
use malva::{config::FormatOptions, format_text};
use raffia::{ast::Stylesheet, Parser, ParserBuilder, Syntax};
use std::fs;

#[test]
fn fmt_snapshot() {
    glob!("fmt/**/*.{css,scss,sass,less}", |path| {
        let input = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };

        let mut comments = vec![];
        let mut parser = ParserBuilder::new(&input)
            .syntax(syntax)
            .comments(&mut comments)
            .build();
        parser
            .parse::<Stylesheet>()
            .map_err(|error| {
                format!(
                    "failed to parse '{}': {} from {} to {}",
                    path.display(),
                    error.kind,
                    error.span.start,
                    error.span.end
                )
            })
            .unwrap();
        let options = if let Some(config) = comments
            .first()
            .and_then(|comment| comment.content.strip_prefix("cfg"))
        {
            toml::from_str::<FormatOptions>(config).unwrap()
        } else {
            Default::default()
        };

        let output = format_text(&input, syntax, &options).unwrap();
        assert!(
            !output.contains(" \n"),
            "'{}' has trailing whitespace",
            path.display()
        );
        let mut regression_parser = Parser::new(&output, syntax);
        assert!(
            regression_parser.parse::<Stylesheet>().is_ok(),
            "'{}' has syntax errors after formatted",
            path.display()
        );

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_snapshot!(name, output);
        });
    });
}
