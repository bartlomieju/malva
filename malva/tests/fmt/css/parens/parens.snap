---
source: malva/tests/fmt.rs
---
a {
  prop1: func(1px, 1px, 1px, func(1px, 1px, 1px, func(1px, 1px, 1px)));
  prop2: func(1px, 1px, 1px, func(1px, 1px, 1px, func(1px, 1px, 1px)));
  prop3: func(1px, 1px, 1px, func(1px, 1px, 1px, func(1px, 1px, 1px)));
  prop4: func(1px, 1px, 1px, func(1px, 1px, 1px, func(1px, 1px, 1px)));
  prop5: func(1px, 1px, 1px, func(1px, 1px, 1px, func(1px, 1px, 1px)));
  prop6: func(1px, 1px, 1px, func(1px, 1px, 1px, func(1px, 1px, 1px)));
  prop7: func(1px, 1px, 1px, func(1px, 1px, 1px, func(1px, 1px, 1px)));
  prop8: very-very-very-very-very-very-vey-very-very-long-func(
    1px,
    1px,
    1px,
    very-very-very-very-very-very-vey-very-very-long-func(
      1px,
      1px,
      1px,
      very-very-very-very-very-very-vey-very-very-long-func(1px, 1px, 1px)
    )
  );
  prop9: very-very-very-very-very-very-vey-very-very-long-func(
    1px,
    1px,
    1px,
    very-very-very-very-very-very-vey-very-very-long-func(
      1px,
      1px,
      1px,
      very-very-very-very-very-very-vey-very-very-long-func(1px, 1px, 1px)
    )
  );
  prop10: very-very-very-very-very-very-vey-very-very-long-func(
    1px,
    1px,
    1px,
    very-very-very-very-very-very-vey-very-very-long-func(
      1px,
      1px,
      1px,
      very-very-very-very-very-very-vey-very-very-long-func(1px, 1px, 1px)
    )
  );
  prop11: very-very-very-very-very-very-vey-very-very-long-func(
    1px,
    1px,
    1px,
    very-very-very-very-very-very-vey-very-very-long-func(
      1px,
      1px,
      1px,
      very-very-very-very-very-very-vey-very-very-long-func(1px, 1px, 1px)
    )
  );
}

.bar {
  prop1: attr(data-foo);
  prop2: attr(data-foo);
  prop3: attr(data-foo);
  prop4: attr(data-foo);
  prop5: attr(data-foo);
  prop6: attr(data-foo);
  prop7: attr(data-foo);
  prop8: attr(data-foo);

  prop9: "(" attr(title) ")";
  prop10: "(" attr(title) ")";
  prop11: "(" attr(title) ")";
  prop12: "(" attr(title) ")";
  prop13: "(" attr(title) ")";
  prop14: "(" attr(title) ")";
  prop15: "(" attr(title) ")";

  prop16: attr(length em);
  prop17: attr(length em);
  prop18: attr(length em);
  prop19: attr(length em);
  prop20: attr(data-size em, 20);
  prop21: attr(data-size em, 20);
  prop22: attr(data-size em, 20);
  prop23: attr(data-size em, 20);
  prop24: attr(data-size em, 20);
  prop25: attr(data-size em, 20);
}

.bar {
  filter: progid:DXImageTransform.Microsoft.gradient(enabled="false",
    startColorstr="#fff", endColorstr="#000");
  filter: progid:DXImageTransform.Microsoft.Shadow(color="#042b47",
    Direction=45, Strength=6)
    progid:DXImageTransform.Microsoft.Shadow(color="#042b47", Direction=135,
    Strength=6);
  -ms-filter: "progid:DXImageTransform.Microsoft.gradient(startColorstr=#fad59f, endColorstr=#fa9907)";
}

.unicode-ranges {
  /* <unicode-range> values */
  unicode-range: U+26; /* single codepoint */
  unicode-range: U+0-7F;
  unicode-range: U+0025-00FF; /* codepoint range */
  unicode-range: U+4??; /* wildcard range */
  unicode-range: U+0025-00FF, U+4??; /* multiple values */
}
